### PR TITLE
feat:(): Flow Metrics aggregation

### PR DESCRIPTION
**Pull request by Skraye**

### CODE REVIEW

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 1]
```diff
@@ -56,4 +58,35 @@ public PagedResults<MetricEntry> findByExecution(
...
+        @Parameter(description = "The namespace") @PathVariable String namespace,
+        @Parameter(description = "The flow Id") @PathVariable String flowId
+    ) {
+        return metricsRepository.flowMetrics(namespace, flowId);
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}/{metric}", produces = MediaType.TEXT_JSON)
+    @Operation(tags = {"Metrics"}, summary = "Get metrics for a specific flow")
+    public ArrayListTotal<MetricStat> aggregateByFlowId(
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:75]: You will not paginate so I think an ArrayList is enought

File: jdbc/src/main/java/io/kestra/jdbc/repository/AbstractJdbcMetricRepository.java [hunk 2]
```diff
@@ -93,6 +160,135 @@ private ArrayListTotal<MetricEntry> query(Condition condition, Pageable pageable
             });
     }
 
+    private List<MetricStat> query(
```
Comment (loicmathieu) [inline jdbc/src/main/java/io/kestra/jdbc/repository/AbstractJdbcMetricRepository.java:163]:
```suggestion
    private List<MetricStat> aggregate(
```

File: core/src/main/java/io/kestra/core/repositories/MetricRepositoryInterface.java [hunk 3]
```diff
@@ -13,6 +16,12 @@ public interface MetricRepositoryInterface extends SaveRepositoryInterface<Metri
 
     ArrayListTotal<MetricEntry> findByExecutionIdAndTaskRunId(String executionId, String taskRunId, Pageable pageable);
 
+    List<String> flowMetrics(String namespace, String flowId);
+
+    ArrayListTotal<MetricStat> aggregateByFlowId(String namespace, String flowId, String metric, ZonedDateTime startDate, ZonedDateTime endDate, String aggregation);
+
+    ArrayListTotal<MetricEntry> aggregateByFlowIdAndTaskId(String flowId, String taskId);
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/repositories/MetricRepositoryInterface.java:23]:
```suggestion
    ArrayListTotal<MetricStat> aggregateByFlowIdAndTaskId(String flowId, String taskId);
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 4]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
             return PagedResults.of(metricsRepository.findByExecutionId(executionId, pageable));
         }
     }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}", produces = MediaType.TEXT_JSON)
+    @Operation(tags = {"Metrics"}, summary = "Get metrics for a specific flow")
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:63]:
```suggestion
    @Operation(tags = {"Metrics"}, summary = "Get metric names for a specific flow")
```

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 5]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+    public List<String> flowMetrics(
+        @Parameter(description = "The namespace") @PathVariable String namespace,
+        @Parameter(description = "The flow Id") @PathVariable String flowId
+    ) {
+        return metricsRepository.flowMetrics(namespace, flowId);
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}/task/{taskId}", produces = MediaType.TEXT_JSON)
+    @Operation(tags = {"Metrics"}, summary = "Get metrics for a specific task in a flow")
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:73]:
```suggestion
    @Operation(tags = {"Metrics"}, summary = "Get metric names for a specific task in a flow")
```

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 6]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+        @Parameter(description = "The namespace") @PathVariable String namespace,
+        @Parameter(description = "The flow Id") @PathVariable String flowId,
+        @Parameter(description = "The flow Id") @PathVariable String taskId
+    ) {
+        return metricsRepository.taskMetrics(namespace, flowId, taskId);
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}/{metric}", produces = MediaType.TEXT_JSON)
+    @Operation(tags = {"Metrics"}, summary = "Get metrics for a specific flow")
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:84]:
```suggestion
    @Operation(tags = {"Metrics"}, summary = "Get metric aggregations for a specific flow")
```

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 7]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+    public MetricAggregations aggregateByFlowId(
+        @Parameter(description = "The namespace") @PathVariable String namespace,
+        @Parameter(description = "The flow Id") @PathVariable String flowId,
+        @Parameter(description = "The metric name") @PathVariable String metric,
+        @Parameter(description = "The start datetime, default to now - 30 days") @Nullable @Format("yyyy-MM-dd'T'HH:mm[:ss][.SSS][XXX]") ZonedDateTime startDate,
+        @Parameter(description = "The end datetime, default to now") @Nullable @Format("yyyy-MM-dd'T'HH:mm[:ss][.SSS][XXX]") ZonedDateTime endDate,
+        @Parameter(description = "The type of aggregation") @QueryValue(defaultValue = "sum") String aggregation
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:91]:
```suggestion
        @Parameter(description = "The type of aggregation: avg, sum, min or max") @QueryValue(defaultValue = "sum") String aggregation
```

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 8]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+            metric,
+            startDate == null ? ZonedDateTime.now().minusDays(30) : startDate,
+            endDate == null ? ZonedDateTime.now() : endDate,
+            aggregation
+        );
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}/task/{taskId}/{metric}", produces = MediaType.TEXT_JSON)
+    @Operation(tags = {"Metrics"}, summary = "Get metrics for a specific flow")
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:106]:
```suggestion
    @Operation(tags = {"Metrics"}, summary = "Get metricagregations for a specific flow")
```

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 9]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+    public MetricAggregations aggregateByFlowIdAndTaskId(
+        @Parameter(description = "The namespace") @PathVariable String namespace,
+        @Parameter(description = "The flow Id") @PathVariable String flowId,
+        @Parameter(description = "The task Id") @PathVariable String taskId,
+        @Parameter(description = "The metric name") @PathVariable String metric,
+        @Parameter(description = "The start datetime, default to now - 30 days") @Nullable @Format("yyyy-MM-dd'T'HH:mm[:ss][.SSS][XXX]") ZonedDateTime startDate,
+        @Parameter(description = "The end datetime, default to now") @Nullable @Format("yyyy-MM-dd'T'HH:mm[:ss][.SSS][XXX]") ZonedDateTime endDate,
+        @Parameter(description = "The type of aggregation") @QueryValue(defaultValue = "sum") String aggregation
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:114]:
```suggestion
        @Parameter(description = "The type of aggregation: avg, sum, min or max") @QueryValue(defaultValue = "sum") String aggregation
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 10]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+    public List<String> taskMetrics(
+        @Parameter(description = "The namespace") @PathVariable String namespace,
+        @Parameter(description = "The flow Id") @PathVariable String flowId,
+        @Parameter(description = "The flow Id") @PathVariable String taskId
+    ) {
+        return metricsRepository.taskMetrics(namespace, flowId, taskId);
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}/{metric}", produces = MediaType.TEXT_JSON)
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:83]:
```suggestion
    @Get(uri = "/aggregates/{namespace}/{flowId}/{metric}", produces = MediaType.TEXT_JSON)
```

File: webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java [hunk 11]
```diff
@@ -56,4 +57,70 @@ public PagedResults<MetricEntry> findByExecution(
...
+            null,
+            metric,
+            startDate == null ? ZonedDateTime.now().minusDays(30) : startDate,
+            endDate == null ? ZonedDateTime.now() : endDate,
+            aggregation
+        );
+    }
+
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "/flow/{namespace}/{flowId}/task/{taskId}/{metric}", produces = MediaType.TEXT_JSON)
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/MetricController.java:105]:
```suggestion
    @Get(uri = "/aggregates/{namespace}/{flowId}/{taskId}/{metric}", produces = MediaType.TEXT_JSON)
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Comment by Skraye** [comment]
close #1226

**Pull request merged**